src/main/java/org/elasticsearch/index/gateway/IndexShardGateway.java

```diff
@@ -157,34 +155,12 @@ public void recover(boolean indexShouldExists, RecoveryState recoveryState) thro
         } finally {
             indexShard.store().decRef();
         }
-        for (final String type : typesToUpdate) {
-            final CountDownLatch latch = new CountDownLatch(1);
-            mappingUpdatedAction.updateMappingOnMaster(indexService.index().name(), indexService.mapperService().documentMapper(type), indexService.indexUUID(), new MappingUpdatedAction.MappingUpdateListener() {
-                @Override
-                public void onMappingUpdate() {
-                    latch.countDown();
-                }
-
-                @Override
-                public void onFailure(Throwable t) {
-                    latch.countDown();
-                    logger.debug("failed to send mapping update post recovery to master for [{}]", t, type);
-                }
-            });
-            cancellableThreads.execute(new CancellableThreads.Interruptable() {
-                @Override
-                public void run() throws InterruptedException {
-                    try {
-                        if (latch.await(waitForMappingUpdatePostRecovery.millis(), TimeUnit.MILLISECONDS) == false) {
-                            logger.debug("waited for mapping update on master for [{}], yet timed out", type);
-                        }
-                    } catch (InterruptedException e) {
-                        logger.debug("interrupted while waiting for mapping update");
-                        throw e;
-                    }
-                }
-            });
-
+        for (Map.Entry<String, Mapping> mappingUpdate : typesToUpdate.entrySet()) {
+            try {
+                MapperUtils.validateDynamicMappingsUpdateOnMaster(client, logger, indexService.index().name(), mappingUpdate.getKey(), mappingUpdate.getValue(), waitForMappingUpdatePostRecovery);
+            } catch (Throwable t) {
+                logger.debug("failed to send mapping update post recovery to master for [{}]", t, mappingUpdate.getKey());
```

Comment:
This seems like a bad case to get into? At least should it be warn or error level? This means we reapplied some updates from the translog, but the master rejected those updates...but we dont seem to do anything with the validation, so does that mean the mappings have already been updated in place? Also, could we limit the catch to just whatever exception would mean failed validation? Otherwise a bug as simple as an NPE in validate would get caught and just logged?